File: packages/astro-identity-gen/src/generate.ts
import { hash, createRng } from "./rng";
import {
  buildPolygonPath,
  edgeStyles,
  type EdgeStyle,
  type PolygonParams,
} from "./polygon";
import { paletteNames, palettes, type PaletteName } from "./theme";
import { buildEyes, eyeStyles, type EyeStyle, type EyeParams } from "./eyes";

export interface IdentityOptions {
  /** Seed string used to deterministically generate the identity. */
  seed: string;
  /** Width/height of the SVG in pixels (default: 128). */
  size?: number;
}

export interface CustomIdentityOptions {
  size?: number;
  bgPalette: PaletteName;
  bgShade: (typeof shadeKeys)[number];
  fgPalette: PaletteName;
  fgShade: (typeof shadeKeys)[number];
  eyePalette: PaletteName;
  eyeShade: (typeof shadeKeys)[number];
  sides: number;
  edgeStyle: EdgeStyle;
  rotation: number;
  radius: number;
  spikeDepth: number;
  curveAmount: number;
  leftEyeStyle: EyeStyle;
  rightEyeStyle: EyeStyle;
  eyeSpacing: number;
  eyeSize: number;
}

/** Structured record of every choice `generateIdentity` makes for a given seed.
 *  Used by the Go port's parity tests to verify identical decisions. */
export interface IdentityChoices {
  size: number;
  bgPalette: PaletteName;
  bgShade: (typeof shadeKeys)[number];
  fgPalette: PaletteName;
  fgShade: (typeof shadeKeys)[number];
  eyePalette: PaletteName;
  eyeShade: (typeof shadeKeys)[number];
  sides: number;
  edgeStyle: EdgeStyle;
  rotation: number;
  radius: number;
  spikeDepth: number;
  curveAmount: number;
  leftEyeStyle: EyeStyle;
  rightEyeStyle: EyeStyle;
  eyeSpacing: number;
  eyeSize: number;
}

/** Pick a value from an array using the next rng output. */
function pick<T>(arr: readonly T[], rng: () => number): T {
  return arr[Math.floor(rng() * arr.length)];
}

/** Pick a value from an array that doesn't match any of the excluded values. */
function pickExcluding<T>(arr: readonly T[], rng: () => number, exclude: T[]): T {
  const filtered = arr.filter((v) => !exclude.includes(v));
  const pool = filtered.length > 0 ? filtered : arr;
  return pool[Math.floor(rng() * pool.length)];
}

/** Shade index in the shadeKeys array. */
function shadeIndex(shade: typeof shadeKeys[number]): number {
  return shadeKeys.indexOf(shade);
}

/** Get shades that are at least `minDistance` steps away from all used shades. */
function shadesWithContrast(
  used: (typeof shadeKeys[number])[],
  minDistance: number,
): (typeof shadeKeys[number])[] {
  return shadeKeys.filter((s) => {
    const si = shadeIndex(s);
    return used.every((u) => Math.abs(si - shadeIndex(u)) >= minDistance);
  });
}

/** Map rng output to a range [min, max]. */
function range(rng: () => number, min: number, max: number): number {
  return min + rng() * (max - min);
}

/** Shade keys ordered light → dark for easy indexing. */
const shadeKeys = [
  50, 100, 200, 300, 400, 500, 600, 700, 800, 900, 950,
] as const;

/**
 * Generates a deterministic SVG identity from a seed string, returning both the
 * SVG and the structured choices the RNG produced. Shared implementation used
 * by `generateIdentity` and parity-fixture tooling.
 */
export function generateIdentityWithChoices(
  options: IdentityOptions,
): { svg: string; choices: IdentityChoices } {
  const { seed, size = 128 } = options;
  const rng = createRng(hash(seed));

  // Pick palettes and shades, ensuring sufficient lightness contrast between layers.
  // minDistance of 3 means at least 3 shade steps apart (~e.g. 200 vs 500).
  const MIN_SHADE_DISTANCE = 3;

  const bgPaletteName = pick(paletteNames, rng);
  const bgShade = pick(shadeKeys, rng);
  const bgColor = palettes[bgPaletteName][bgShade];

  const fgPaletteName = pick(paletteNames, rng);
  const fgPool = shadesWithContrast([bgShade], MIN_SHADE_DISTANCE);
  const fgShade = fgPool.length > 0 ? pick(fgPool, rng) : pickExcluding(shadeKeys, rng, [bgShade]);
  const fgColor = palettes[fgPaletteName][fgShade];

  // Derive polygon parameters
  const sides = Math.floor(range(rng, 3, 9)); // 3–8
  const edgeStyle: EdgeStyle = pick(edgeStyles, rng);
  const rotation = range(rng, 0, 2 * Math.PI);
  const radius = range(rng, 0.55, 0.9);
  const spikeDepth = range(rng, 0.3, 0.7);
  const curveAmount = range(rng, 0.2, 0.5);

  const params: PolygonParams = {
    sides,
    edgeStyle,
    rotation,
    radius,
    spikeDepth,
    curveAmount,
  };

  const path = buildPolygonPath(params, size);

  // Eyes — ensure sufficient lightness contrast against both bg and polygon
  const eyePaletteName = pick(paletteNames, rng);
  const eyePool = shadesWithContrast([bgShade, fgShade], MIN_SHADE_DISTANCE);
  const eyeShade = eyePool.length > 0 ? pick(eyePool, rng) : pickExcluding(shadeKeys, rng, [bgShade, fgShade]);
  const eyeColor = palettes[eyePaletteName][eyeShade];
  const leftEyeStyle = pick(eyeStyles, rng);
  // ~10% chance of mismatched eyes
  const rightEyeStyle = rng() < 0.1 ? pickExcluding(eyeStyles, rng, [leftEyeStyle]) : leftEyeStyle;
  const eyeSpacing = range(rng, 0.15, 0.35);
  const eyeSize = range(rng, 0.04, 0.1);
  const eyeParams: EyeParams = {
    leftStyle: leftEyeStyle,
    rightStyle: rightEyeStyle,
    spacing: eyeSpacing,
    eyeSize,
  };
  const eyes = buildEyes(eyeParams, size, eyeColor);

  const svg = [
    `<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 ${size} ${size}">`,
    `  <rect width="${size}" height="${size}" fill="${bgColor}" />`,
    `  <path d="${path}" fill="${fgColor}" />`,
    `  ${eyes}`,
    `</svg>`,
  ].join("\n");

  const choices: IdentityChoices = {
    size,
    bgPalette: bgPaletteName,
    bgShade,
    fgPalette: fgPaletteName,
    fgShade,
    eyePalette: eyePaletteName,
    eyeShade,
    sides,
    edgeStyle,
    rotation,
    radius,
    spikeDepth,
    curveAmount,
    leftEyeStyle,
    rightEyeStyle,
    eyeSpacing,
    eyeSize,
  };

  return { svg, choices };
}

/**
 * Generates a deterministic SVG identity from a seed string.
 */
export function generateIdentity(options: IdentityOptions): string {
  return generateIdentityWithChoices(options).svg;
}

/**
 * Generates an SVG identity from explicit trait values (no seed / RNG).
 */
export function generateCustomIdentity(opts: CustomIdentityOptions): string {
  const size = opts.size ?? 128;

  const bgColor = palettes[opts.bgPalette][opts.bgShade];
  const fgColor = palettes[opts.fgPalette][opts.fgShade];
  const eyeColor = palettes[opts.eyePalette][opts.eyeShade];

  const params: PolygonParams = {
    sides: opts.sides,
    edgeStyle: opts.edgeStyle,
    rotation: opts.rotation,
    radius: opts.radius,
    spikeDepth: opts.spikeDepth,
    curveAmount: opts.curveAmount,
  };

  const path = buildPolygonPath(params, size);

  const eyeParams: EyeParams = {
    leftStyle: opts.leftEyeStyle,
    rightStyle: opts.rightEyeStyle,
    spacing: opts.eyeSpacing,
    eyeSize: opts.eyeSize,
  };
  const eyes = buildEyes(eyeParams, size, eyeColor);

  return [
    `<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 ${size} ${size}">`,
    `  <rect width="${size}" height="${size}" fill="${bgColor}" />`,
    `  <path d="${path}" fill="${fgColor}" />`,
    `  ${eyes}`,
    `</svg>`,
  ].join("\n");
}

File: apps/astro-server/internal/identitygen/generate.go
package identitygen

import (
	"fmt"
	"math"
	"slices"
	"strings"
)

// IdentityOptions controls seeded identity generation.
type IdentityOptions struct {
	Seed string
	Size int // 0 means default 128
}

// CustomIdentityOptions specifies every identity trait explicitly, bypassing RNG.
type CustomIdentityOptions struct {
	Size          int
	BGPalette     string
	BGShade       int
	FGPalette     string
	FGShade       int
	EyePalette    string
	EyeShade      int
	Sides         int
	EdgeStyle     EdgeStyle
	Rotation      float64
	Radius        float64
	SpikeDepth    float64
	CurveAmount   float64
	LeftEyeStyle  EyeStyle
	RightEyeStyle EyeStyle
	EyeSpacing    float64
	EyeSize       float64
}

// identityChoices is the structured record of every decision generateIdentity
// makes for a given seed. Used by parity tests.
type identityChoices struct {
	Size          int
	BGPalette     string
	BGShade       int
	FGPalette     string
	FGShade       int
	EyePalette    string
	EyeShade      int
	Sides         int
	EdgeStyle     EdgeStyle
	Rotation      float64
	Radius        float64
	SpikeDepth    float64
	CurveAmount   float64
	LeftEyeStyle  EyeStyle
	RightEyeStyle EyeStyle
	EyeSpacing    float64
	EyeSize       float64
}

const minShadeDistance = 3

// GenerateIdentity returns a deterministic SVG identity for the given seed.
func GenerateIdentity(opts IdentityOptions) string {
	svg, _ := generateIdentityWithChoices(opts)
	return svg
}

// GenerateCustomIdentity returns an SVG built from explicit traits (no RNG).
func GenerateCustomIdentity(opts CustomIdentityOptions) string {
	size := opts.Size
	if size == 0 {
		size = 128
	}

	bgColor := palettes[opts.BGPalette][opts.BGShade]
	fgColor := palettes[opts.FGPalette][opts.FGShade]
	eyeColor := palettes[opts.EyePalette][opts.EyeShade]

	path := buildPolygonPath(PolygonParams{
		Sides:       opts.Sides,
		EdgeStyle:   opts.EdgeStyle,
		Rotation:    opts.Rotation,
		Radius:      opts.Radius,
		SpikeDepth:  opts.SpikeDepth,
		CurveAmount: opts.CurveAmount,
	}, size)

	eyes := buildEyes(EyeParams{
		LeftStyle:  opts.LeftEyeStyle,
		RightStyle: opts.RightEyeStyle,
		Spacing:    opts.EyeSpacing,
		EyeSize:    opts.EyeSize,
	}, size, eyeColor)

	return assembleSVG(size, bgColor, fgColor, path, eyes)
}

// generateIdentityWithChoices is the internal implementation shared by
// GenerateIdentity and tests. Every rng() call must occur in the same sequence
// as the TypeScript reference — that sequence is the parity contract.
func generateIdentityWithChoices(opts IdentityOptions) (string, identityChoices) {
	size := opts.Size
	if size == 0 {
		size = 128
	}

	rng := createRng(hash(opts.Seed))

	bgPalette := pick(paletteNames, rng)
	bgShade := pick(shadeKeys, rng)
	bgColor := palettes[bgPalette][bgShade]

	fgPalette := pick(paletteNames, rng)
	fgPool := shadesWithContrast([]int{bgShade}, minShadeDistance)
	var fgShade int
	if len(fgPool) > 0 {
		fgShade = pick(fgPool, rng)
	} else {
		fgShade = pickExcluding(shadeKeys, rng, []int{bgShade})
	}
	fgColor := palettes[fgPalette][fgShade]

	sides := int(math.Floor(rangeFloat(rng, 3, 9))) // 3–8
	edgeStyle := pick(edgeStyles, rng)
	rotation := rangeFloat(rng, 0, 2*math.Pi)
	radius := rangeFloat(rng, 0.55, 0.9)
	spikeDepth := rangeFloat(rng, 0.3, 0.7)
	curveAmount := rangeFloat(rng, 0.2, 0.5)

	path := buildPolygonPath(PolygonParams{
		Sides:       sides,
		EdgeStyle:   edgeStyle,
		Rotation:    rotation,
		Radius:      radius,
		SpikeDepth:  spikeDepth,
		CurveAmount: curveAmount,
	}, size)

	eyePalette := pick(paletteNames, rng)
	eyePool := shadesWithContrast([]int{bgShade, fgShade}, minShadeDistance)
	var eyeShade int
	if len(eyePool) > 0 {
		eyeShade = pick(eyePool, rng)
	} else {
		eyeShade = pickExcluding(shadeKeys, rng, []int{bgShade, fgShade})
	}
	eyeColor := palettes[eyePalette][eyeShade]

	leftEyeStyle := pick(eyeStyles, rng)
	// ~10% chance of mismatched eyes; the rng() call happens unconditionally here.
	var rightEyeStyle EyeStyle
	if rng() < 0.1 {
		rightEyeStyle = pickExcluding(eyeStyles, rng, []EyeStyle{leftEyeStyle})
	} else {
		rightEyeStyle = leftEyeStyle
	}
	eyeSpacing := rangeFloat(rng, 0.15, 0.35)
	eyeSize := rangeFloat(rng, 0.04, 0.1)

	eyes := buildEyes(EyeParams{
		LeftStyle:  leftEyeStyle,
		RightStyle: rightEyeStyle,
		Spacing:    eyeSpacing,
		EyeSize:    eyeSize,
	}, size, eyeColor)

	choices := identityChoices{
		Size:          size,
		BGPalette:     bgPalette,
		BGShade:       bgShade,
		FGPalette:     fgPalette,
		FGShade:       fgShade,
		EyePalette:    eyePalette,
		EyeShade:      eyeShade,
		Sides:         sides,
		EdgeStyle:     edgeStyle,
		Rotation:      rotation,
		Radius:        radius,
		SpikeDepth:    spikeDepth,
		CurveAmount:   curveAmount,
		LeftEyeStyle:  leftEyeStyle,
		RightEyeStyle: rightEyeStyle,
		EyeSpacing:    eyeSpacing,
		EyeSize:       eyeSize,
	}
	return assembleSVG(size, bgColor, fgColor, path, eyes), choices
}

func assembleSVG(size int, bgColor, fgColor, path, eyes string) string {
	var b strings.Builder
	fmt.Fprintf(&b, `<svg xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" viewBox="0 0 %d %d">`, size, size, size, size)
	b.WriteByte('\n')
	fmt.Fprintf(&b, `  <rect width="%d" height="%d" fill="%s" />`, size, size, bgColor)
	b.WriteByte('\n')
	fmt.Fprintf(&b, `  <path d="%s" fill="%s" />`, path, fgColor)
	b.WriteByte('\n')
	b.WriteString("  ")
	b.WriteString(eyes)
	b.WriteByte('\n')
	b.WriteString(`</svg>`)
	return b.String()
}

// pick selects an element from arr using the next rng output, mirroring
// `arr[Math.floor(rng() * arr.length)]`.
func pick[T any](arr []T, rng func() float64) T {
	return arr[int(math.Floor(rng()*float64(len(arr))))]
}

// pickExcluding is pick, but drawing from (arr - exclude); if that's empty it
// falls back to arr so behavior matches the TS reference.
func pickExcluding[T comparable](arr []T, rng func() float64, exclude []T) T {
	filtered := make([]T, 0, len(arr))
	for _, v := range arr {
		if !slices.Contains(exclude, v) {
			filtered = append(filtered, v)
		}
	}
	pool := filtered
	if len(pool) == 0 {
		pool = arr
	}
	return pool[int(math.Floor(rng()*float64(len(pool))))]
}

// shadesWithContrast returns the shades at least `minDistance` steps away from
// every shade in `used` in the shadeKeys ordering.
func shadesWithContrast(used []int, minDistance int) []int {
	out := make([]int, 0, len(shadeKeys))
	for _, s := range shadeKeys {
		si := slices.Index(shadeKeys, s)
		ok := true
		for _, u := range used {
			if abs(si-slices.Index(shadeKeys, u)) < minDistance {
				ok = false
				break
			}
		}
		if ok {
			out = append(out, s)
		}
	}
	return out
}

// rangeFloat maps a rng() output to [min, max).
func rangeFloat(rng func() float64, minV, maxV float64) float64 {
	return minV + rng()*(maxV-minV)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
